packages/services/src/activity/stacks-tx-activity.utils.ts

```diff
@@ -0,0 +1,194 @@
+import {
+  ContractCallTransaction,
+  MempoolContractCallTransaction,
+  MempoolSmartContractTransaction,
+  MempoolTokenTransferTransaction,
+  SmartContractTransaction,
+  TokenTransferTransaction,
+  TransactionEvent,
+} from '@stacks/stacks-blockchain-api-types';
+
+import { stxCryptoAsset } from '@leather.io/constants';
+import {
+  AccountIdentifier,
+  ActivityLevels,
+  OnChainActivity,
+  OnChainActivityStatuses,
+  OnChainActivityTypes,
+  ReceiveAssetActivity,
+  SendAssetActivity,
+} from '@leather.io/models';
+import { initBigNumber, uniqueArray } from '@leather.io/utils';
+
+import {
+  HiroStacksMempoolTransaction,
+  HiroStacksTransaction,
+  HiroTransactionEvent,
+} from '../infrastructure/api/hiro/hiro-stacks-api.client';
+import {
+  StacksAssetTransferWithInfo,
+  aggregateTransferReceivers,
+  aggregateTransferSenders,
+  sumAssetTransferAmounts,
+} from './stacks-asset-transfer.utils';
+
+export function getEventsByTxId(events: TransactionEvent[]) {
+  return events.reduce((acc, event) => {
+    if (!acc.has(event.tx_id)) {
+      acc.set(event.tx_id, []);
+    }
+    acc.get(event.tx_id)?.push(event);
+    return acc;
+  }, new Map<string, HiroTransactionEvent[]>());
+}
+
+export function mapStacksTxBlockTime(tx: HiroStacksTransaction | HiroStacksMempoolTransaction) {
+  if (isMempoolTx(tx)) {
+    return tx.receipt_time;
+  } else {
+    return (tx as any).block_time ?? tx.burn_block_time;
+  }
+}
+
+export function isMempoolTx(
+  tx: HiroStacksTransaction | HiroStacksMempoolTransaction
+): tx is HiroStacksMempoolTransaction {
+  return (
+    tx.tx_status !== 'success' &&
+    tx.tx_status !== 'abort_by_post_condition' &&
+    tx.tx_status !== 'abort_by_response'
+  );
+}
```

Comment:
pretty sure we have a similar method to this in stacks package? If not, could this live there as a pure fn?